### ISSUE TITLE
Interaction with Dependency Injector

### ISSUE THREAD

**Issue by YannickAlex07** [microsoft/pyright#4785]
**Describe the bug**

Hey there! We use Pyright together with the [dependency injector](https://github.com/ets-labs/python-dependency-injector) (`4.41.0`) package and the newest Pyright version (`1.1.299`) is giving is some type issues with it.

We are not sure if the issue is a bug in Pyright or rather an issue in the injector package, however because the update to Pyright caused the issues - and we can't see a point in the changelog that sounds like it relates to the issue - we open the issue here.

The issue we are facing seems related to generic classes and metaclasses. Injecting dependencies works as follows:

```python
from dependency_injector import containers, providers
from dependency_injector.wiring import Provide, inject

class Container(containers.DeclarativeContainer):
    my_class = providers.Singleton(
        MyClass,
    )

@inject
def my_function(
    my_dependency: MyClass = Provide[Container.my_class],
):
    pass
```

Prior to `1.1.299`, this caused no issues, however since the update the function parameter (`Provide[...]`) is giving us the following errors:

* `Expected no type arguments for class "Provide"`
* `Type[ClassGetItemMeta]" is incompatible with "Type[MyClass]`

The package is annotating the `Provide` as follows ([here is the code](https://github.com/ets-labs/python-dependency-injector/blob/cc2304e46e054ae08dc12995428759fbfb51af10/src/dependency_injector/wiring.py#L849-L873)):

```python
class ClassGetItemMeta(GenericMeta):
    def __getitem__(cls, item):
        ...

class _Marker(Generic[T], metaclass=ClassGetItemMeta):
    ...

class Provide(_Marker):
    ...
```

Again, we don't know - and we lack on typing knowledge here - if this is an issue in Pyright or in the dependency injector package. It would be nice if this could be checked, so that we know what to do.

**To Reproduce**

Any example from the [package website](https://python-dependency-injector.ets-labs.org/index.html) can be used to reproduce the issue. Furthermore the code above can also be used of course.

**Expected behavior**

The expected behavior is that no error message should show for the provided code

**VS Code extension or command-line**

The newest VSCode extension is not causing this issue - however it seems like it is still using `1.1.298`. Running via the CLI with version `1.1.299` will show the errors.

**Comment by erictraut** [comment]
This is a bug in the library. The class `Provide` is not generic, so it doesn't accept any generic type arguments. The expression `Provide[Container.my_class]` can be evaluated at runtime because the metaclass used for the `Provide` class supplies a `__getitem__` method, but that doens't make `Provide` generic.

There was previously a bug in pyright where it silently ignored type arguments to non-generic classes when the metaclass supplied a `__getitem__` method. This bug was fixed in 1.1.299.

The correct fix here is for `Provide` to use a type variable in its class definition, as follows:

```python
# Current definition of `Provide` omits the type argument for `_Marker`, so
# it is properly interpreted as `_Marker[Any]`.
class Provide(_Marker):
    ...

# Correct (generic) definition of `Provide`:
class Provide(_Marker[T]):
    ...
```

Please report the bug to the library maintainer.

**Comment by erictraut** [comment]
This will be addressed in the next release of pyright.

**Comment by YannickAlex07** [comment]
Thanks a lot @erictraut!

**Comment by erictraut** [comment]
This is included in pyright 1.1.300, which I just published. It will also be included in a future release of pylance.
